### ISSUE TITLE
services.udev.extraHwdb rules that worked on NixOS 20.09 no longer work on 21.05

### ISSUE THREAD

**Issue by kini** [NixOS/nixpkgs#127443]
**Describe the bug**

My services.udev.extraHwdb rules for swapping keyboard keys worked on NixOS 20.09 but no longer work on 21.05.

**To Reproduce**

I have the following in my `configuration.nix`:

```nix
    # keyboard swaps
    services.udev.extraHwdb = ''
      evdev:input:b*v04F2p0833*  # USB keyboard: AmazonBasics wired ("CHICONY USB Keyboard")
      evdev:input:b*v045Ep07A5*  # USB keyboard: Microsoft Ergonomics Sculpt wireless ("Microsoft Microsoft® 2.4GHz Transceiver v9.0")
       KEYBOARD_KEY_70029=capslock  # map esc to caps lock
       KEYBOARD_KEY_70039=esc       # map caps lock to esc
       KEYBOARD_KEY_700e6=rightctrl # map right alt to right ctrl
       KEYBOARD_KEY_700e4=rightalt  # map right ctrl to right alt
    '';
```

This works fine on NixOS 20.09, i.e. I observe that the esc and caps lock keys are swapped and the right alt and right ctrl keys are swapped. But on NixOS 21.05, the rules don't seem to be doing anything.

I inspected the /etc/udev/hwdb.bin files from a 20.09 system profile and a 21.05 system profile on my machine in a hex editor. I'm not sure what the data format of those files is, but I can see that last few bytes of both files seem to contain text from my `services.udev.extraHwdb` setting. Before:

```
008b4930: 7030 0034 6570 3132 3231 3a2a 006e 7075  p0.4ep1221:*.npu
008b4940: 743a 6200 7630 3446 3270 3038 3333 2a00  t:b.v04F2p0833*.
008b4950: 204b 4559 424f 4152 445f 4b45 595f 3730   KEYBOARD_KEY_70
008b4960: 3032 3900 7630 3400 4570 3037 4135 2a00  029.v04.Ep07A5*.
008b4970: 204b 4559 424f 4152 445f 4b45 595f 3730   KEYBOARD_KEY_70
008b4980: 3033 3900 204b 4559 424f 4152 445f 4b45  039. KEYBOARD_KE
008b4990: 595f 3730 3065 3600 7269 6768 7463 7472  Y_700e6.rightctr
008b49a0: 6c00 204b 4559 424f 4152 445f 4b45 595f  l. KEYBOARD_KEY_
008b49b0: 3730 3065 3400 7269 6768 7461 6c74 00    700e4.rightalt.
```

After:

```
008d2440: 0034 6570 3132 3231 3a2a 0061 2a00 6e70  .4ep1221:*.a*.np
008d2450: 7574 3a62 0076 3034 4632 7030 3833 332a  ut:b.v04F2p0833*
008d2460: 0020 4b45 5942 4f41 5244 5f4b 4559 5f37  . KEYBOARD_KEY_7
008d2470: 3030 3239 0076 3034 0045 7030 3741 352a  0029.v04.Ep07A5*
008d2480: 0020 4b45 5942 4f41 5244 5f4b 4559 5f37  . KEYBOARD_KEY_7
008d2490: 3030 3339 0020 4b45 5942 4f41 5244 5f4b  0039. KEYBOARD_K
008d24a0: 4559 5f37 3030 6536 0072 6967 6874 6374  EY_700e6.rightct
008d24b0: 726c 0020 4b45 5942 4f41 5244 5f4b 4559  rl. KEYBOARD_KEY
008d24c0: 5f37 3030 6534 0072 6967 6874 616c 7400  _700e4.rightalt.
```

So presumably whatever changed to cause the breakage was on the reading end of `hwdb.bin` rather than the writing end. Not sure how to investigate any further.

(BTW, I see that `udevadm hwdb` has been deprecated in favor of `systemd-hwdb`, so maybe we should change the command that generates the `hwdb.bin` file regardless - but that's a separate matter.)

**Expected behavior**

I expected the esc and caps lock keys to be swapped and the right alt and right ctrl keys to be swapped on my Microsoft Ergonomics Sculpt wireless keyboard after booting into a system with the described value of the `services.udev.extraHwdb` NixOS option.

**Notify maintainers**
There don't seem to be any maintainers of the udev module, so I guess I'll just ping the last few people from the git history of the module who seem to have made possibly relevant changes to the module: @rnhmjoj @flokli @shlevy. Sorry if you weren't the right people to ping.

**Metadata**
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.43, NixOS, 21.05.1076.bad3ccd099e (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.12`
 - channels(root): `"nixos-20.09.4321.115dbbe82eb"`
 - channels(fs): `"nixpkgs-21.11pre293448.2a1c29ef4ba"`
 - nixpkgs: `/home/fs/.nix-defexpr/channels/nixpkgs`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

**Comment by rnhmjoj** [comment]
Sorry, I don't know much about hwdb. If you just need to swap a few keys, a possible workaround is to use xkb. You can design a layout based on your current one but with keys swapped and install it via `services.xserver.extraLayouts`. If you also add `console.useXkbConfig` you have remapped the keys in virtually all applications.

There is an example in the manual if you want to try: https://nixos.org/manual/nixos/stable/#custom-xkb-layouts.

**Comment by kini** [comment]
Thanks for the response. However, if possible I would prefer to continue to use a more low-level modification based on hwdb, because I've had bad experiences with higher-level ones (e.g. `xmodmap`) which don't seem to interact well with Chinese/Japanese/Korean input method editor software, for example, and with desktop environments like KDE which have their own management of keyboard layouts in X. Doing these swaps at the udev level ensures that nothing higher level can mess with or even notice the swaps.

**Comment by rnhmjoj** [comment]
If we could automate a check "keys have been swapped" it would be easy to run a bisect and find the culprit.
`xorg.xev` could probably do it, I don't know if the `hwdb` method works in a qemu VM, though.

**Comment by rnhmjoj** [comment]
Ok, I just on 20.09, 21.05 and master and It worked every time.
This is what I used to test your setup:

```nix
import nixos/tests/make-test-python.nix ({...} :

{
  name = "swapping-keyboard-keys";

  machine = { pkgs, ... }:
    {
      documentation.enable = false;
      users.users.root.password = "";
      services.udev.extraHwdb =
        ''
          evdev:name:AT Translated Set 2 keyboard:*
           KEYBOARD_KEY_2c=x  # z -> x
           KEYBOARD_KEY_2d=z  # x -> z
        '';
    };

  testScript =
    ''
      machine.start()
      machine.wait_for_unit("getty.target")
      machine.sleep(2)
      machine.send_chars("root\n")
      machine.sleep(2)
      machine.send_key("ret")
      machine.sleep(2)
      machine.send_chars("echo zx > /test\n")
      machine.wait_for_file("/test")
      machine.succeed("grep -q xz /test")
    '';
})
```

My guess would be the identifier of you keyboard has changed in between
kernel/udev updates. I don't what method you used to get those ids
(`v04F2p0833`, etc), but I checked the name of the device with
```
grep Name= /proc/bus/input/devices
```
and then used something like
```
evdev:name:AT Translated Set 2 keyboard:*
```
so, maybe try this, instead.

**Comment by kini** [comment]
It turns out this was caused by PEBKAC :facepalm: . Very sorry for the noise! I'll close this issue now. Read further if you're interested in what exactly the problem was...

After some more poking around I found that the settings are actually being applied correctly in a VT (as your test shows), and failing only in the graphical environment. (This was true even with my particular keyboard identifiers based on `evdev:input:...`, so that wasn't the problem.)

Digging further, I found that immediately after running `udevadm trigger` in a terminal, the key swaps appeared to be in effect, but as soon as I switched focus to another window or tab (my terminal emulator, Konsole, has a tabbed interface), the key swaps seemed to be getting undone by something, and they didn't come back until I ran `udevadm trigger` again.

Hunting through `journalctl -f` I happened to notice some warning messages from `xmodmap` (!) if I happened to be holding a modifier key like Shift or Alt for more than a couple seconds while changing focus between windows or tabs right after running `udevadm trigger`...

Then I realized there was an ancient `~/.Xmodmap` file in my homedir, which I used to use ages ago to perform the same swaps that I'm now doing with hwdb. Apparently something has changed in some component of my desktop environment in 21.05 (I guess some component of KDE) so that xmodmap settings are cleaned and reloaded from `~/.Xmodmap` every time focus is changed, or something. (Could also be an interaction with my Japanese input method editor, which has the ability to change input contexts on window refocus, though I have that turned off.) In 20.09, afaik that file was never loaded automatically, and even years ago when I was actually using it, I had to manually trigger it by running `xmodmap` in a terminal...

And of course, since the key swaps are self-inverting, applying them twice (once from the hwdb rules and once from xmodmap) would cancel out... :facepalm:

Sorry again for the noise!
